File: 210-course-schedule-ii/course-schedule-ii.py
class Solution(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        indegree = defaultdict(int)

        for pair in prerequisites:
            indegree[pair[0]] += 1

        queue = deque(course for course in range(numCourses) if indegree[course] == 0)

        result_order = []
        while queue:
            current_course = queue.popleft()
            result_order.append(current_course)

            for pair in prerequisites:
                if pair[1] == current_course:
                    indegree[pair[0]] -= 1
                    if indegree[pair[0]] == 0:
                        queue.append(pair[0])

        if len(result_order) == numCourses:
            return result_order
        else:
            return []

File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];

        for (int[] pair : prerequisites) {
            indegree[pair[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        for (int course = 0; course < numCourses; course++) {
            if (indegree[course] == 0) {
                queue.offer(course);
            }
        }

        List<Integer> resultOrder = new ArrayList<>();
        while (!queue.isEmpty()) {
            int currentCourse = queue.poll();
            resultOrder.add(currentCourse);

            for (int[] pair : prerequisites) {
                if (pair[1] == currentCourse) {
                    indegree[pair[0]]--;
                    if (indegree[pair[0]] == 0) {
                        queue.offer(pair[0]);
                    }
                }
            }
        }

        if (resultOrder.size() == numCourses) {
            int[] resultArray = new int[numCourses];
            for (int i = 0; i < numCourses; i++) {
                resultArray[i] = resultOrder.get(i);
            }
            return resultArray;
        } else {
            return new int[]{};
        }
    }
}